ledger/participant-state/kvutils/app/src/test/scala/com/daml/ledger/participant/state/kvutils/app/RunnerSpec.scala

```diff
@@ -0,0 +1,264 @@
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AsyncWordSpec
+
+import scala.collection.mutable
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.existentials
+
+class RunnerSpec extends AsyncWordSpec with Matchers {
+  private implicit val resourceContext: ResourceContext = ResourceContext(ExecutionContext.global)
+  private implicit val loggingContext: LoggingContext = LoggingContext.ForTesting
+
+  "the runner" should {
+    "start a participant" in {
+      val config = Config.createDefault(()).copy(ledgerId = LedgerId)
+      val participantConfig = newTestPartcipantConfig()
+      val runner = new Runner(Name, TestLedgerFactory, ConfigProvider.ForUnit)
+
+      newApp(config, participantConfig, runner).use { channel =>
+        LedgerIdentityServiceGrpc
+          .stub(channel)
+          .getLedgerIdentity(GetLedgerIdentityRequest.of())
+          .map { response =>
+            response.ledgerId should be(LedgerId)
+          }
+      }
+    }
+  }
+}
+
+object RunnerSpec {
+  private val Name = classOf[RunnerSpec].getSimpleName
+  private val LedgerId = s"$Name-Ledger"
+
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private val engine = Engine.StableEngine()
+
+  private def newTestPartcipantConfig() = {
+    val participantId = Ref.ParticipantId.assertFromString("participant")
+    ParticipantConfig(
+      mode = ParticipantRunMode.Combined,
+      participantId = participantId,
+      shardName = None,
+      address = None,
+      port = Port.Dynamic,
+      portFile = None,
+      serverJdbcUrl = ParticipantConfig.defaultIndexJdbcUrl(participantId),
+      indexerConfig = ParticipantIndexerConfig(allowExistingSchema = false),
+    )
+  }
+
+  private def newApp[T <: ReadWriteService](
+      config: Config[Unit],
+      participantConfig: ParticipantConfig,
+      runner: Runner[T, Unit],
+  )(implicit
+      loggingContext: LoggingContext
+  ): ResourceOwner[Channel] =
+    for {
+      actorSystem <- ResourceOwner.forActorSystem(() => ActorSystem(Name))
+      materializer <- ResourceOwner.forMaterializer(() => Materializer(actorSystem))
+      port <- new ResourceOwner[Port] {
+        override def acquire()(implicit context: ResourceContext): Resource[Port] =
+          runner
+            .runParticipant(config, participantConfig, engine)(
+              context,
+              loggingContext,
+              actorSystem,
+              materializer,
+            )
+            .map(_.get)
+      }
+      channel <- ResourceOwner
+        .forChannel(
+          {
+            val builder = ManagedChannelBuilder
+              .forAddress(InetAddress.getLoopbackAddress.getHostName, port.value)
+            builder.usePlaintext()
+            builder
+          },
+          shutdownTimeout = 1.second,
+        )
+    } yield channel
+
+  object TestLedgerFactory extends LedgerFactory[Unit] {
```

Comment:
This seems like a lot of overhead considering that we are really interested only in the healthchecks. Wouldn't a mock work better?

Reply:
The API server won't start until we get a configuration out. We also need to handle the submission fairly normally otherwise we get errors upon submitting an initial ledger configuration. While this doesn't actually make the test fail, it does result in a big stack trace in the logs that will probably make it harder to figure out what's going wrong if/when the test fails in the future.